src/EditorFeatures/Core/Shared/Utilities/ResettableDelay.cs

```diff
@@ -9,6 +9,8 @@ namespace Microsoft.CodeAnalysis.Editor.Shared.Utilities
 {
     internal class ResettableDelay
     {
+        public static readonly ResettableDelay CompletedDelay = new ResettableDelay();
```

Comment:
this needs documentation, or a better name.  It reads really weird to have a ResettableDelay.CompletedDelay.  Specifically, you are exposing a ResettableDelay that can't actually be reset and which will not ever call a timer.  So it's both not something to reset, and it's not a delay.  So the name itself is incredibly strange and indicates something wonky.

Reply:
> . so not sure what you mean by it is weird that it won't reset once it is completed

The type is literally called "ResetableDelay", but you're exposing an instance that is neither resettable (since reset won't actually do anything), nor does it have a delay :)

Task.CompletedTask is not the same situation as Task is not a mutable entity.   the closest thing here is TaskCompletionSource, and yes, i would find it very weird if TaskCompletionSource exposed a static, completed, task (or source). :)

Reply:
I think this is one of those good reasons to not reinvent the wheel.  The semantics of task/tcs have been well established since their introduction almost 10 years ago (it was VS2010 right?).  And async/await is well understood and common throughout our codebase.  By using all the stock tools we already have, it helps avoid confusion and makes hte code clear for a wider audience.  By going with bespoke solutions, we make the code less idiomatic and harder to understand (see my comments elsewhere in this PR about trying to wrap my head around things, and trying to figure out if things are correct).

Is there a valuable reason to be different than the rest of the patterns we use in our codebase?  Do we need to avoid existing tools that seem more than sufficient *and much clearer* for this sort of problem space?

Reply:
well. again, I am fine to create tracking issue to remove ResettableDelay time from code base. but won't remove it in this PR or not use it.

Reply:
tracking issue - https://github.com/dotnet/roslyn/issues/26397

Reply:
I have not asked you to remove ResettableDelay from the codebase.  I've just poitned out that you've used it here in a seemingly unnecessary and overly complex manner.  it seems like for the purposes of hte change you want to make here, there are simpler and clearer patterns that can be used that still accomplish what you are trying to accomplish.  As part of a PR i'm always going to point out if i think there are substantively better approaches that can be taken that improve the health of the code and make things easier to PR.

Reply:
I used it as it is used in other places. if we have that type, no reason not to use it. if it is not reasonable to be used here, then it shouldn't be reasonable to be used other places as well. 

I dont see any of your argument not applied to other places if we accept your arguments.

Reply:
>  if we have that type, no reason not to use it. 

There is a reason: it's less idiomatic and more bespoke than existing patterns nad practices already provided by the language and libraries.  I'm making not comment about the rest of the code, i'm specifically talking about *this* pr.  Based on the change you are trying to make here, i'm not certain i see the value in this type.  I odn't want to hitch these decision to a larger decision about "what should we do elsewhere".  It's quite possible in the rest of the code that that type is perfectly fine and sufficient.  However, that may be hte case for those types.  In this type, i haven't seen a good reason why it is necessary, especially as it seems much more complex than what we could accomplish with existing platform/language offerings.

Reply:
well. I accept the type as one of type we accept to use. like AsyncLazy, ValueSource or any other helper type we created. I can argue why I need to understand AsyncLazy, ValueSource, instead of just using SemaphoreSlim.DisposableWaitAsync. 

same goes for our complex tagger rather than simple ITagger from editor and etc. but I accept that is one of thing we have so I learn how it works and use it.

ResettableDelay is there, so I learned how it works and I am using it as it is supposed to be used.

Reply:
and I am not sure why you are saying it is not existing pattern. it is existing pattern.

Reply:
and I dont see it any more complex than using cancellation in tagger, background parser and etc.

Reply:
> I can argue why I need to understand AsyncLazy,

AsyncLazy has specific reasons it exists taht are borne out from the actual framework types being insufficient.  That said, if you feel like it can be dropped/simplifie,d then i would absolutely support that if it does not regress anything.  But it was not done on a whim.  it was done because our attempt to just use the normal framework types turned out to be insufficient.

> ValueSource

I would be 100% getting rid of ValueSource if possible.  I do not see why the presence of other possibly unnecessary parts of the system is a good justificaiton for making *more* unnecessary parts.

> and I dont see it any more complex than using cancellation in tagger, background parser and etc.

You are literally hand rolling tasks instead of using async/await.  You are coming up with a complex cancellation and delay scheme instead of just using cancellation tokens and Task.Delay. These are literal ways of reinventing what can already be done in a simple and built-in fashion in the language and platform.  

Now, reinvention can be appropriate *if* the existing systems are insufficient or have some sort of drawback.  That's precisely why we've written our own primitives at time.  

Also, some things were written this way because they *predate* async/await and hte ability to easily write async code in C# itself.  If we were to write those systems *today* we would likely do things a bit differently.

But this code of yours is *new* code in a very modern component.  We only added the editor_config stuff in the last couple of years.  There is no reason in these components to avoid just using the bog standard framework/language tools that are already at our disposable and are not only more than sufficient, but they're well understood and commonly used across the entire codebase.

--

Put it htis way:  every time you write a custom solution, that's an extra cognitive burden for anyone who has to understand and maintain this code.  Instead of having the same patterns used across the rest of the codebase, there are bespoke patterns used *almost nowhere*.  I looked at ResettableDelay.  It's in *2* classes GCmanager and GraphQueryManager, whereas tasks, cancellation tokens, and sources, are used in literally thousands of places.

Instead of using the code that is much more common and well understood, you used what looks to be an unnecessary type even though there has been no justification provided for why it would be superior for this use case over the standard patterns.

To write well maintainable code, things should be consistent and should involve doing things in a special-cased fashion as *few times* as possible, and only when *necessary*.  If it is necessary (i.e. a good reason can be given), then that's a reason to special-case things.  Otherwise the code is adding complexity, and maintenance cost for no good reason at all.  And that's just not good for the health of hte codebase.

Reply:
> and I dont see it any more complex than using cancellation in tagger, background parser and etc.

You are welcome to simplify any of those systems if you want.  No one will stop you .  If you can move to more idiomatic C# patterns, and you can replace code with framework code, taht would be *fantastic*.

Much of that code was written *before* c# even had async/await.  So it contains things that were felt to be necessary at the time, but may not be now.  We should not use the approaches we took there, under hte **constraints** we had at that time, to indicate that we should also be side-stepping hte frameworks and language for *new* code that we are writing.  

Note: i would be totally willing to improve cancellaiton in tagging.  Can you be specific in terms of what you thing needs improvement?

Reply:
Note: i took a peek at how the assembly checking logic works.  It looks like it could potentially be made into something that doesn't not need both langauges to be from the same type. Merely that it needs be able to get the public keys of an assembly (no way to do that now publicly afaict), and then check them against the 'Identity' of the assembly (which is public).

So it feels like assembly checking can be done in a language-boundary agnostic fashion.  Though i will admit i didn't look exhaustively to verify this.